src/core/Ekubo.sol

```diff
@@ -0,0 +1,453 @@
+        uint256 skipAhead
+    ) external payable returns (int128 delta0, int128 delta1);
+
+    // Pay for swapped tokens
+    function pay(address token) external returns (uint128 payment);
+
+    // Get swapped tokens
+    function withdraw(address token, address recipient, uint128 amount) external;
+}
+
+IEkuboCore constant CORE = IEkuboCore(0xe0e0e08A6A4b9Dc7bD67BCB7aadE5cF48157d444);
+
+/// @notice Interface for the callback executed when an address locks core
+interface IEkuboCallbacks {
+    /// @notice Called by Core on `msg.sender` when a lock is acquired
+    /// @param id The id assigned to the action
+    /// @return Any data that you want to be returned from the lock call
+    function locked(uint256 id) external returns (bytes memory);
+
+    /// @notice Called by Core on `msg.sender` to collect assets
+    /// @param id The id assigned to the action
+    /// @param token The token to pay on
+    function payCallback(uint256 id, address token) external;
+}
+
+library UnsafeEkuboCore {
+    function unsafeSwap(
+        IEkuboCore core,
+        PoolKey memory poolKey,
+        int128 amount,
+        bool isToken1,
+        SqrtRatio sqrtRatioLimit,
+        uint256 skipAhead
+    ) internal returns (int128 delta0, int128 delta1) {
+        assembly ("memory-safe") {
+            let ptr := mload(0x40)
+
+            mstore(ptr, 0x00000000) // selector for `swap_611415377((address,address,bytes32),int128,bool,uint96,uint256)`
+            // TODO: Check that native ETH is handled properly (0x0 vs 0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee)
+            mcopy(add(0x20, ptr), poolKey, 0x60)
+            mstore(add(0x80, ptr), signextend(0x0f, amount))
+            mstore(add(0xa0, ptr), isToken1)
+            mstore(add(0xc0, ptr), and(0xffffffffffffffffffffffff, sqrtRatioLimit))
+            mstore(add(0xe0, ptr), skipAhead)
+            
+            if iszero(call(gas(), core, 0x00, add(0x1c, ptr), 0xe4, 0x00, 0x40)) {
+                returndatacopy(ptr, 0x00, returndatasize())
+                revert(ptr, returndatasize())
+            }
+            delta0 := mload(0x00)
+            delta1 := mload(0x20)
+        }
+    }
+}
+
+abstract contract Ekubo is SettlerAbstract {
+    using UnsafeMath for uint256;
+    using UnsafeMath for int256;
+    using CreditDebt for int256;
+    using Ternary for bool;
+    using SafeTransferLib for IERC20;
+    using NotesLib for NotesLib.Note;
+    using NotesLib for NotesLib.Note[];
+    using StateLib for StateLib.State;
+    using UnsafeEkuboCore for IEkuboCore;
+
+    constructor() {
+        assert(BASIS == Encoder.BASIS);
+        assert(BASIS == Decoder.BASIS);
+        assert(ETH_ADDRESS == Decoder.ETH_ADDRESS);
+    }
+
+    // fill encoding musings:
+    //     bps (2 bytes)
+    //     packing key (1 byte)
+    //     sell/buy tokens (0, 20, or 40 bytes)
+    //     config (32 bytes) - (8 bytes fee, 4 bytes tickSpacing, 20 bytes extension)
+    //     skipAhead (32 bytes)
```

Comment:
There's no possibility for hookdata with Ekubo? It's not something clever where the hookdata is appended to the end of the call to `swap_611415377`?

Reply:
Extensions (which are the "hooks" in Ekubo) don't receive any special data from the call, just some of the known parameters/data in use. So, it doesn't seem to be possible to have any hookdata.
<img width="735" alt="image" src="https://github.com/user-attachments/assets/409bf225-138f-4be0-8465-9e7582ffaf47" />
<img width="632" alt="image" src="https://github.com/user-attachments/assets/4c727652-6036-46af-a246-c8fe220c739a" />
Extensions docs [here](https://docs.ekubo.org/integration-guides/extensions)